### PR TITLE
Add precise filtering of which items to translate

### PR DESCRIPTION
**Pull request by Nadrieril**
This PR adds a rust implementation of something like the `NameMatcher`. This makes it possible to precisely filter items precisely, e.g. translate exactly one impl from the standard library. This fixes the issue of `--extract-opaque-bodies` triggering a slurry of errors. This is still a bit rough but already useful.

Fixes https://github.com/AeneasVerif/charon/issues/214

### CODE REVIEW

File: charon-ml/src/LlbcOfJson.ml [hunk 1]
```diff
@@ -181,43 +181,42 @@ let split_global (gid_conv : global_id_converter) global :
   (global_decl, fun_decl)
 
 let crate_of_json (js : json) : (crate, string) result =
-  combine_error_msgs js __FUNCTION__
```
Comment (sonmarcho) [inline charon-ml/src/LlbcOfJson.ml:184]: The `combine_error_msgs` is not used anymore?

File: charon/src/name_matcher/mod.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+enum PatElem {
+    /// An identifier, optionally with generic arguments. E.g. `std` or `Box<_>`.
+    Ident {
+        name: String,
+        generics: Vec<PatTy>,
+        /// For pretty-printing only: whether this is the name of a trait.
+        is_trait: bool,
```
Comment (sonmarcho) [inline charon/src/name_matcher/mod.rs:24]: Is this only used for pretty printing, or also for matching?
For the Charon-ML side: we should have the same, but should also use it for matching (if `true`, we can only match a trait impl).

**Comment by sonmarcho** [comment]
It's super nice to finally have this, thanks :)

File: charon-ml/src/LlbcOfJson.ml (hunk 1 above)
Comment (Nadrieril) [reply charon-ml/src/LlbcOfJson.ml:184]: This is only for the version mismatch error: we don't want to dump the parsed AST in that case.

File: charon/src/name_matcher/mod.rs (hunk 2 above)
Comment (Nadrieril) [reply charon/src/name_matcher/mod.rs:24]: I'm not sure how useful that is: there cannot be a trait with the same name as a type. The fact that we ignore this means we can support both the `impl Trait<..> for Self` syntax and the `Trait<Self, ..>` syntaxes.

Comment (Nadrieril) [reply charon/src/name_matcher/mod.rs:24]: Well, I guess with wildcards there can be :thinking:

Comment (sonmarcho) [reply charon/src/name_matcher/mod.rs:24]: Actually you're probably right: it may be useless. But I was mostly thinking about consistency/sanity...

**Pull request merged**